### PR TITLE
Registration prequal component

### PR DESCRIPTION
**Pull request by MocanaAtBroad**
**Update May 19**
Applied refactoring suggested by @fat23cat.
Included invitation code formatting directive that is fully operational now.

**Older message**
Just the basics with the invitation code, the zip code and the captcha button.
All 3 are needed to submit.
I have wired it to the register button.
No styling to speak of. 
Was working on directive to provide nice formatting of invitation code, but keep getting into problems with implementation.
Problems are fixable, but don't want to delay getting making component available to have something to work with.

### CODE REVIEW

File: ddp-workspace/package.json [hunk 1]
```diff
@@ -35,6 +35,7 @@
     "core-js": "2.5.4",
     "hammerjs": "2.0.8",
     "kaop-ts": "3.1.2",
+    "ng-recaptcha": "^5.0.0",
```
Comment (MocanaAtBroad) [inline ddp-workspace/package.json]: Picked up this third party library. Works well and is properly documented.

File: ddp-workspace/projects/ddp-sdk/src/lib/components/age-up/verifyAgeUp.component.ts [hunk 2]
```diff
@@ -14,7 +13,7 @@ export class VerifyAgeUpComponent implements OnInit {
 
     constructor(
         private route: ActivatedRoute,
-        private invitationsService: InvitationsServiceAgent) { }
+        private invitationsService: InvitationServiceAgent) { }
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/age-up/verifyAgeUp.component.ts:16]: Made name singular. More common.

File: ddp-workspace/projects/ddp-sdk/src/lib/invitationCodeFormatter.directive.ts [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+        {
+            provide: NG_VALUE_ACCESSOR,
+            useExisting: forwardRef(() => InvitationCodeFormatterDirective),
+            multi: true,
+        }
+    ]
+})
+
+export class InvitationCodeFormatterDirective implements ControlValueAccessor {
+    private _value: string | null = '';
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/invitationCodeFormatter.directive.ts]: This is the directive that is not quite working right yet, so don't waste too much time looking at it.
Supposed to take in the entered string and format it by uppercasing and introducing separators every 4 characters.
That part is easy.
The part that is harder is to deal updating the insertion point in the text box after someone has typed or pasted text from anywhere in the input box.
Again, have not resolved all the issues there, but at least I understand them.

File: ddp-workspace/projects/ddp-sdk/src/lib/services/serviceAgents/invitationServiceAgent.service.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+            invitationId,
+            zip,
+            recaptchaToken,
+            auth0ClientId: this.configuration.auth0ClientId};
+        return this.postObservable(`/studies/${this.configuration.studyGuid}/invitation-check`,  payload, {}, true)
+            .pipe(
+                map(response => response ? response.body : null),
+                catchError(err => {
+                    const errorType = err.status === 400 ? ErrorType.InvalidInvitation : ErrorType.Unknown;
+                    return throwError(new DdpError('Problem with submission', errorType));
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/services/serviceAgents/invitationServiceAgent.service.ts]: Wanted to make sure we treated 400 as validation errors for which we would alert user.

File: ddp-workspace/projects/ddp-sdk/src/lib/services/serviceAgents/invitationsServiceAgent.service.ts [hunk 5]
```diff
@@ -1,26 +0,0 @@
-import { Injectable, Inject } from '@angular/core';
-import { HttpClient } from '@angular/common/http';
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/services/serviceAgents/invitationsServiceAgent.service.ts:2]: Gone with the rename

File: ddp-workspace/projects/ddp-testboston/src/app/app-routes.ts [hunk 6]
```diff
@@ -9,5 +11,6 @@ export const AppRoutes = {
     CovidSurvey: 'covid-survey',
     Dashboard: 'dashboard',
     Password: 'password',
-    SessionExpired: 'session-expired'
+    SessionExpired: 'session-expired',
+    UserRegistrationPrequal: 'user-registration-prequal'
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-testboston/src/app/app-routes.ts]: The path for the component.I believe that we want to have component live in home page. Doable but punting on that for now just to make it easier to work with and get to, I give its own route.

File: ddp-workspace/projects/ddp-testboston/src/app/app-routing.module.ts [hunk 7]
```diff
@@ -109,6 +110,13 @@ const routes: Routes = [
       IrbGuard
     ]
   },
+  {
+    path: AppRoutes.UserRegistrationPrequal,
+    component: UserRegistrationPrequalComponent,
+    canActivate: [
+      IrbGuard
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-testboston/src/app/app-routing.module.ts]: Note there is no authentication filter

File: ddp-workspace/projects/ddp-testboston/src/app/app-routes.ts (hunk 6 above)
Comment (fat23cat) [reply ddp-workspace/projects/ddp-testboston/src/app/app-routes.ts]: As I remember, Jen asked to use `join` route name for this page.

File: ddp-workspace/projects/ddp-testboston/src/app/app.module.ts [hunk 8]
```diff
@@ -51,6 +55,7 @@ toolkitConfig.covidSurveyGuid = AppGuids.Covid;
 toolkitConfig.dashboardGuid = AppGuids.Dashboard;
 toolkitConfig.phone = 'XXX-XXX-XXXX';
 toolkitConfig.infoEmail = 'testboston@datadonationplatform.org';
+toolkitConfig.recaptchaSiteKey = '6LdqYvUUAAAAAFl_KZFyNQBT3dMjvVnTb-P9wfAs';
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-testboston/src/app/app.module.ts:67]: Needs to be part of configuration file. Coming soon.

File: ddp-workspace/projects/ddp-testboston/src/app/app.module.ts [hunk 9]
```diff
@@ -69,63 +74,70 @@ sdkConfig.auth0Audience = DDP_ENV.auth0Audience;
...
+            const locale = 'en';
+            translate.setDefaultLang(locale);
+            translate.use(locale).subscribe(() => {
+                console.log(`Successfully initialized '${locale}' language as default.`);
+            }, err => {
+                console.error(`Problem with '${locale}' language initialization.`);
+            }, () => {
+                resolve(null);
+            });
+        });
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-testboston/src/app/app.module.ts]: Just code reformatting.

File: ddp-workspace/projects/ddp-testboston/src/app/app.module.ts [hunk 10]
```diff
@@ -69,63 +74,70 @@ sdkConfig.auth0Audience = DDP_ENV.auth0Audience;
...
+            provide: APP_INITIALIZER,
+            useFactory: translateFactory,
+            deps: [
+                TranslateService,
+                Injector
+            ],
+            multi: true
+        }
+    ],
+    bootstrap: [AppComponent]
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-testboston/src/app/app.module.ts]: Just code reformatting. Not sure why this happened, but no code was changed here.

File: ddp-workspace/projects/ddp-testboston/src/app/app.module.ts [hunk 11]
```diff
@@ -69,63 +74,70 @@ sdkConfig.auth0Audience = DDP_ENV.auth0Audience;
...
-      ga('send', event);
-      ga('platform.send', event);
-    });
-  }
+    constructor(private analytics: AnalyticsEventsService) {
+        this.analytics.analyticEvents.subscribe((event: AnalyticsEvent) => {
+            ga('send', event);
+            ga('platform.send', event);
+        });
+    }
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-testboston/src/app/app.module.ts]: Code formatting.

File: ddp-workspace/projects/ddp-testboston/src/app/models/errorType.ts [hunk 12]
```diff
@@ -0,0 +1,4 @@
+export enum ErrorType {
+    InvalidInvitation = 'InvalidInvitation',
+    Unknown = 'Unknown'
+}
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-testboston/src/app/models/errorType.ts]: Defining some error types. This might change as the back-end/apis are updated.

File: ddp-workspace/projects/ddp-testboston/src/app/app-routes.ts (hunk 6 above)
Comment (fat23cat) [reply ddp-workspace/projects/ddp-testboston/src/app/app-routes.ts]: For the reference https://broadinstitute.slack.com/archives/G012Y2S45KN/p1589484057371700

File: ddp-workspace/projects/ddp-testboston/src/app/models/ddpError.ts [hunk 13]
```diff
@@ -0,0 +1,9 @@
+import { ErrorType } from './errorType';
+
+export class DdpError extends Error {
+    errorType: ErrorType;
+    constructor(msg: string, errorType: ErrorType) {
+        super(msg);
+        this.errorType = errorType;
+    }
+}
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-testboston/src/app/models/ddpError.ts]: Surprised we have not had one of these before. Well here it is. I am using with new component.

File: ddp-workspace/projects/ddp-testboston/src/app/user-registration-prequal/userRegistrationPrequal.component.ts [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+        <re-captcha [siteKey]="config.recaptchaSiteKey" size="normal" formControlName="recaptchaToken"></re-captcha>
+        <button type="submit" [disabled]="formGroup.invalid" mat-raised-button color="primary">Submit</button>
+      </form>
+  `,
+  styles: [`
+      .registration-prequal-form {
+        display: flex;
+        flex-direction: column;
+      }
+  `]
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-testboston/src/app/user-registration-prequal/userRegistrationPrequal.component.ts]: Need to come back and internationalize.

File: ddp-workspace/projects/ddp-testboston/src/app/user-registration-prequal/userRegistrationPrequal.component.ts [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+          <mat-label>Enter invitation code</mat-label>
+          <input matInput
+                 uppercase
+                 formControlName="invitationId"
+                 pattern="[a-zA-Z0-9]{12}"
+                 maxLength="12"
+                 placeholder="Your invitation code">
+        </mat-form-field>
+        <mat-form-field class="example-input-field" floatLabel="always">
+          <mat-label>What is your current mailing zip code?</mat-label>
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-testboston/src/app/user-registration-prequal/userRegistrationPrequal.component.ts]: Need to use plain material widget but tried to pickup whatever styling class we were using.

File: ddp-workspace/projects/toolkit/src/lib/services/toolkitConfiguration.service.ts [hunk 16]
```diff
@@ -62,4 +62,7 @@ export class ToolkitConfigurationService {
     showInfoForPhysicians: boolean;
     showBlog: boolean;
     blogUrl: string;
+
+    // security
+    recaptchaSiteKey: string;
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/toolkit/src/lib/services/toolkitConfiguration.service.ts:67]: Adding to toolkit as I expect that this will become a part of future apps

File: ddp-workspace/tslint.json [hunk 17]
```diff
@@ -183,18 +183,16 @@
     "directive-selector": [
       true,
       "attribute",
-      "app",
       "camelCase"
     ],
     "component-selector": [
       true,
       "element",
-      "app",
```
Comment (MocanaAtBroad) [inline ddp-workspace/tslint.json:192]: the 'components selectors need to start with `app` prefix' was annoying long enough. Same for directives.

File: ddp-workspace/projects/ddp-testboston/src/app/app-routes.ts (hunk 6 above)
Comment (MocanaAtBroad) [reply ddp-workspace/projects/ddp-testboston/src/app/app-routes.ts]: 385e0df

File: ddp-workspace/projects/ddp-sdk/src/lib/services/serviceAgents/invitationServiceAgent.service.ts [hunk 18]
```diff
@@ -0,0 +1,48 @@
+import { Inject, Injectable } from '@angular/core';
+import { HttpClient, HttpErrorResponse } from '@angular/common/http';
+import { NotAuthenticatedServiceAgent } from './notAuthenticatedServiceAgent.service';
+import { ConfigurationService } from '../configuration.service';
+import { LoggingService } from '../logging.service';
+import { EMPTY, Observable, throwError } from 'rxjs';
+import { catchError, map, mergeMap } from 'rxjs/operators';
+import { InvitationCheckPayload } from '../../../../../ddp-testboston/src/app/models/invitationCheckPayload';
+import { DdpError } from '../../../../../ddp-testboston/src/app/models/ddpError';
+import { ErrorType } from '../../../../../ddp-testboston/src/app/models/errorType';
```
Comment (fat23cat) [inline ddp-workspace/projects/ddp-sdk/src/lib/services/serviceAgents/invitationServiceAgent.service.ts]: It looks like `InvitationCheckPayload`, `DdpError`, `ErrorType` should be part of the SDK, we shouldn't import anything from the app to the library.

File: ddp-workspace/projects/ddp-sdk/src/public-api.ts [hunk 19]
```diff
@@ -63,10 +63,12 @@ export * from './lib/services/renewSessionNotifier.service';
 export * from './lib/services/scriptLoader.service';
 export * from './lib/services/serviceAgents/governedParticipantsServiceAgent.service';
 export * from './lib/services/languageService.service';
+export * from './lib/services/serviceAgents/invitationServiceAgent.service';
 
 export * from './lib/components/login/auth0-code-callback.component';
 export * from './lib/components/address/addressEmbedded.component';
 export * from './lib/components/activityForm/activity.component';
+export * from './lib/directives/upperCaseInputDirective.directive';
```
Comment (fat23cat) [inline ddp-workspace/projects/ddp-sdk/src/public-api.ts]: I moved upperCaseInputDirective in the directives folder.

File: ddp-workspace/projects/ddp-testboston/src/app/app-routes.ts [hunk 20]
```diff
@@ -3,11 +3,11 @@ export const AppRoutes = {
     ActivityId: 'activity/:id',
     ActivityLinkId: 'activity-link/:id',
     Error: 'error',
-    Prequalifier: 'prequalifier',
```
Comment (fat23cat) [inline ddp-workspace/projects/ddp-testboston/src/app/app-routes.ts]: Removed unused route const.

File: ddp-workspace/projects/ddp-testboston/src/app/app-routing.module.ts [hunk 21]
```diff
@@ -23,13 +23,6 @@ import {
 } from 'toolkit';
 
 const routes: Routes = [
-  {
-    path: AppRoutes.Prequalifier,
-    component: WorkflowStartActivityRedesignedComponent,
-    canActivate: [
-      IrbGuard
-    ]
-  },
```
Comment (fat23cat) [inline ddp-workspace/projects/ddp-testboston/src/app/app-routing.module.ts]: Instead of `WorkflowStartActivityRedesignedComponent` we use `UserRegistrationPrequalComponent`

File: ddp-workspace/projects/ddp-testboston/src/app/components/user-registration-prequal/user-registration-prequal.component.ts [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+import { FormControl, FormGroup, Validators } from '@angular/forms';
+import { ToolkitConfigurationService } from 'toolkit';
+import { Auth0AdapterService, InvitationServiceAgent } from 'ddp-sdk';
+import { ErrorType } from '../../models/errorType';
+import { take } from 'rxjs/operators';
+
+@Component({
+  selector: 'user-registration-prequal',
+  templateUrl: './user-registration-prequal.component.html',
+  styleUrls: ['./user-registration-prequal.component.scss']
```
Comment (fat23cat) [inline ddp-workspace/projects/ddp-testboston/src/app/components/user-registration-prequal/user-registration-prequal.component.ts:10]: I split component into html and scss files, because it's extremely convenient to use html and scss tips and formatting.

File: ddp-workspace/projects/ddp-testboston/src/app/components/user-registration-prequal/user-registration-prequal.component.ts [hunk 23]
```diff
@@ -0,0 +1,51 @@
...
+  public errorMessage: string | null = null;
+
+  constructor(
+    private auth0: Auth0AdapterService,
+    private invitationService: InvitationServiceAgent,
+    @Inject('toolkit.toolkitConfig') public config: ToolkitConfigurationService) { }
+
+  public ngOnInit(): void {
+    this.initForm();
+  }
```
Comment (fat23cat) [inline ddp-workspace/projects/ddp-testboston/src/app/components/user-registration-prequal/user-registration-prequal.component.ts:23]: I extracted form initialization from the constructor, primarily, we use constructor only for injecting dependencies.

File: ddp-workspace/projects/ddp-testboston/src/app/models/ddpError.ts (hunk 13 above)
Comment (fat23cat) [reply ddp-workspace/projects/ddp-testboston/src/app/models/ddpError.ts]: Cool!

File: ddp-workspace/projects/ddp-sdk/src/public-api.ts [hunk 24]
```diff
@@ -63,10 +63,12 @@ export * from './lib/services/renewSessionNotifier.service';
 export * from './lib/services/scriptLoader.service';
 export * from './lib/services/serviceAgents/governedParticipantsServiceAgent.service';
 export * from './lib/services/languageService.service';
+export * from './lib/services/serviceAgents/invitationServiceAgent.service';
```
Comment (fat23cat) [inline ddp-workspace/projects/ddp-sdk/src/public-api.ts:66]: If we use a service outside of the library, we should export it.

**Comment by fat23cat** [comment]
So, I've updated template and styles, it looks a bit better:
image
I didn't add fonts and other styles, because most of them are now in #112

File: ddp-workspace/projects/ddp-sdk/src/lib/directives/invitationCodeFormatter.directive.ts [hunk 25]
```diff
@@ -0,0 +1,98 @@
+import { Directive, ElementRef, forwardRef, HostListener, Input } from '@angular/core';
+import { MAT_INPUT_VALUE_ACCESSOR } from '@angular/material';
+import { ControlValueAccessor, NG_VALUE_ACCESSOR } from '@angular/forms';
+import { InvitationCodeInputFormatter } from '../utility/invitationCode/invitationCodeInputFormatter';
+
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/directives/invitationCodeFormatter.directive.ts:5]: Rewrote directive delegating all the formatting and positioning of cursor to invitationCodeInputFormtter

File: ddp-workspace/projects/ddp-sdk/src/lib/directives/invitationCodeFormatter.directive.ts [hunk 26]
```diff
@@ -0,0 +1,98 @@
...
+        value = value == null ? '' : value;
+        console.debug('set value was called with value: %o', value);
+        const cleanedInputState = this.formatter.cleanupInput(value, this.elementRef.nativeElement.selectionStart);
+        this.setModelValue(cleanedInputState.value);
+        const formattedState = this.formatter.addSeparator(cleanedInputState.value, cleanedInputState.selectionStart, this.isBackspace);
+        this.elementRef.nativeElement.value = formattedState.value;
+        this.elementRef.nativeElement.selectionStart = formattedState.selectionStart;
+        this.elementRef.nativeElement.selectionEnd = formattedState.selectionStart;
+    }
+
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/directives/invitationCodeFormatter.directive.ts:43]: All the other stuff in directive is to make Angular happy etc. We take the provided in put and the selection start todo all calculations. Special case when backspace is pressed. We pick it below.

File: ddp-workspace/projects/ddp-sdk/src/lib/models/errorType.ts [hunk 27]
```diff
@@ -0,0 +1,4 @@
+export enum ErrorType {
+    InvalidInvitation = 'INVALID_INVITATION',
+    InvalidQualifications = 'INVALID_INVITATION_QUALIFICATIONS'
+}
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/models/errorType.ts:4]: Changed these to match codes on server.

File: ddp-workspace/projects/ddp-sdk/src/lib/models/textInputState.ts [hunk 28]
```diff
@@ -0,0 +1,5 @@
+export interface TextInputState {
+    value: string;
+    selectionStart: number;
+    isBackSpace?: boolean;
+}
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/models/textInputState.ts:5]: Support model class for directive. Capture everything we need to know about state to calculate display and position of cursor.

File: ddp-workspace/projects/ddp-sdk/src/lib/services/serviceAgents/invitationServiceAgent.service.ts [hunk 29]
```diff
@@ -0,0 +1,49 @@
...
+            recaptchaToken,
+            auth0ClientId: this.configuration.auth0ClientId
+        };
+        return this.postObservable(`/studies/${this.configuration.studyGuid}/invitation-check`, payload, {}, true)
+            .pipe(
+                map(response => response ? response.body : null),
+                catchError((err: HttpErrorResponse) => {
+                    return throwError(this.buildErrorObject(err));
+                })
+            );
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/services/serviceAgents/invitationServiceAgent.service.ts:36]: Other stuff was here before. This is what we call for user registration.

File: ddp-workspace/projects/ddp-sdk/src/lib/utility/invitationCode/invitationCodeInputFormatter.spec.ts [hunk 30]
```diff
@@ -0,0 +1,93 @@
...
+    expect(formatterObject.filter('1 - X8', 4)).toEqual({ value: '1X8', selectionStart: 1 });
+  });
+
+  it('format and addding chunk from border', () => {
+    expect(formatterObject.addSeparator('1234ZA56', 5)).toEqual({ value: '1234 - ZA56 - ', selectionStart: 8 });
+  });
+  it('format after hitting backspace on last character', () => {
+    expect(formatterObject.addSeparator('XB28LQ8PZNH', 11, true)).toEqual({ value: 'XB28 - LQ8P - ZNH', selectionStart: 17 });
+  });
+
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/utility/invitationCode/invitationCodeInputFormatter.spec.ts:83]: A lot of weird situations to evaluate in formatter.

File: ddp-workspace/projects/ddp-sdk/src/lib/utility/invitationCode/invitationCodeInputFormatter.ts [hunk 31]
```diff
@@ -0,0 +1,75 @@
...
+        return { value: formattedString, selectionStart };
+    }
+
+    format = (inputState: TextInputState): TextInputState => {
+        let currentState = this.cleanupInput(inputState.value, inputState.selectionStart);
+        currentState = this.addSeparator(currentState.value, currentState.selectionStart, inputState.isBackSpace);
+        console.debug('the state after addSeparator is %o', currentState);
+        return currentState;
+    }
+
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/utility/invitationCode/invitationCodeInputFormatter.ts:53]: Easiest to see main idea here: cleanInput takes whatever is provided in input box and tries to come up with valid code as expected by the server (convert to uppercase, toss out non-alpha chars, etc.).
Note that important trick here is to manage the cursor insertion point to be in right spot as we add or remove characters during processing. Otherwise behavior in widget gets weird when typing or pasting content.
separator takes the cleaned-out input and lays it out, and once again tries to position cursor "property".

File: ddp-workspace/projects/ddp-testboston/src/app/components/user-registration-prequal/user-registration-prequal.component.html [hunk 32]
```diff
@@ -0,0 +1,38 @@
...
+    <section class="section">
+        <div class="content">
+            <h1>First, enter your invitation code and zip code</h1>
+            <p>If you received a paper or email invitation from your doctor,
+                it includes a 12-digit invitation code at the bottom.</p>
+            <form class="invitation-form" [formGroup]="formGroup" (ngSubmit)="onSubmit()">
+                <mat-form-field class="invitation-form__input" floatLabel="always">
+                    <mat-label>Enter invitation code</mat-label>
+                    <input matInput
+                           invitationcode
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-testboston/src/app/components/user-registration-prequal/user-registration-prequal.component.html:11]: Directive selector added back in.

File: ddp-workspace/projects/ddp-testboston/src/app/components/user-registration-prequal/user-registration-prequal.component.scss [hunk 33]
```diff
@@ -0,0 +1,59 @@
...
+        }
+    }
+}
+
+.ErrorMessage {
+    margin: 1rem 0;
+    color: #ee4416;
+    font-size: 1.375rem;
+    line-height: 2.0625rem;
+}
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-testboston/src/app/components/user-registration-prequal/user-registration-prequal.component.scss:59]: Styling looking good.

File: ddp-workspace/projects/ddp-testboston/src/app/components/user-registration-prequal/user-registration-prequal.component.ts (hunk 22 above)
Comment (MocanaAtBroad) [reply ddp-workspace/projects/ddp-testboston/src/app/components/user-registration-prequal/user-registration-prequal.component.ts:10]: Sure! I personally prefer to keep it all in one file as I find it easier to keep it all together, but IDE does not really help me much, so happy to split things up.

**Review by MocanaAtBroad — COMMENTED** [review]
Thanks for changes @fat23cat . I have incorporated your changes and moved things around to follow your advice. The big thing now is that directive entry works as I originally intended.

File: ddp-workspace/projects/ddp-testboston/src/app/components/user-registration-prequal/user-registration-prequal.component.html [hunk 34]
```diff
@@ -0,0 +1,38 @@
+<main class="main">
+    <section class="section">
+        <div class="content">
+            <h1>First, enter your invitation code and zip code</h1>
+            <p>If you received a paper or email invitation from your doctor,
+                it includes a 12-digit invitation code at the bottom.</p>
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-testboston/src/app/components/user-registration-prequal/user-registration-prequal.component.html:6]: @fat23cat Text you added said 16 digits. We have implemented 12 digits. Is Jen under impression we doing 16. Not a big deal but better if no surprises.

File: ddp-workspace/projects/ddp-sdk/src/lib/directives/invitationCodeFormatter.directive.ts [hunk 35]
```diff
@@ -0,0 +1,98 @@
+import { Directive, ElementRef, forwardRef, HostListener, Input } from '@angular/core';
+import { MAT_INPUT_VALUE_ACCESSOR } from '@angular/material';
+import { ControlValueAccessor, NG_VALUE_ACCESSOR } from '@angular/forms';
+import { InvitationCodeInputFormatter } from '../utility/invitationCode/invitationCodeInputFormatter';
+
+@Directive({
+    selector: 'input[invitationcode]',
+    providers: [
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/directives/invitationCodeFormatter.directive.ts:8]: Will filter garbage input and try to always present in text box a format XXXX - XXXX - XXXX while at the same time updating the model to present the API -expected 12 character code.
Backspacing, pasting values with our without dashes, typing by hand,etc. etc. all are supposed to work here.

File: ddp-workspace/projects/ddp-testboston/src/app/components/user-registration-prequal/user-registration-prequal.component.html [hunk 36]
```diff
@@ -0,0 +1,38 @@
...
+                </mat-form-field>
+                <mat-form-field class="invitation-form__input" floatLabel="always">
+                    <mat-label>What is your current mailing zip code?</mat-label>
+                    <input matInput
+                           formControlName="zip"
+                           minlength="5"
+                           maxlength="5"
+                           placeholder="5 digit zip"
+                           type="number"
+                           pattern="[0-9]{5}">
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-testboston/src/app/components/user-registration-prequal/user-registration-prequal.component.html:24]: Had to take out type="number". Will not validate otherwise.

File: ddp-workspace/projects/ddp-sdk/src/public-api.ts (hunk 19 above)
Comment (fat23cat) [reply ddp-workspace/projects/ddp-sdk/src/public-api.ts]: @MocanaAtBroad, why do you think that's a bad idea?

File: ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/picklist/checkboxesActivityPicklistQuestion.component.spec.ts [hunk 37]
```diff
@@ -8,138 +8,4 @@ import { MatListModule } from '@angular/material/list';
...
-
-        it('should add detais text to answer', () => {
-            fixture.detectChanges();
-            component.select(true, 'AAA', true);
-            component.detailTextChanged('Text', 'AAA');
-            expect(component.block.answer).toEqual([{ stableId: 'AAA', detail: 'Text' }]);
-        });
-    });
-});
+describe('CheckboxesActivityPicklistQuestion', () => {});
```
Comment (fat23cat) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/picklist/checkboxesActivityPicklistQuestion.component.spec.ts:11]: Why did you remove this test?

**Comment by fat23cat** [comment]
Do we have any list of valid invitation codes or util for generating invitation code? Maybe we need something like this for testing user registration?
cc @zyme, @yufengwng

File: ddp-workspace/projects/ddp-sdk/src/public-api.ts (hunk 19 above)
Comment (fat23cat) [reply ddp-workspace/projects/ddp-sdk/src/public-api.ts]: Ah, we also can remove this export, it is not necessary, importing in the SDK module is enough.

File: ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/picklist/checkboxesActivityPicklistQuestion.component.spec.ts (hunk 37 above)
Comment (MocanaAtBroad) [reply ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/picklist/checkboxesActivityPicklistQuestion.component.spec.ts:11]: Test is broken/does not compile. And when that happens you can't run any other tests.
All the tests are fixed in the mail address refactoring branch and has circleci running the tests. We should merge it soon as I don't expect to do any further work there.

**Comment by MocanaAtBroad** [comment]
> Do we have any list of valid invitation codes or util for generating invitation code? Maybe we need something like this for testing user registration?
> cc @zyme, @yufengwng
There is a valid code in the test. I also I generated a bunch into the dev database. I'll share them with team via Google Docs or something.

File: ddp-workspace/projects/ddp-sdk/src/public-api.ts (hunk 19 above)
Comment (MocanaAtBroad) [reply ddp-workspace/projects/ddp-sdk/src/public-api.ts]: Sorry! I meant to give you a thumbs up! 👍 
Actually, I think I can remove it now as it is no longer used now that we have directive that formats the code exactly as needed, including upper casing.
It is likely that the invitation code input box will become its own components as we want to use same idea in future projects and there will be no need to export the directives.

File: ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/picklist/checkboxesActivityPicklistQuestion.component.spec.ts (hunk 37 above)
Comment (fat23cat) [reply ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/picklist/checkboxesActivityPicklistQuestion.component.spec.ts:11]: If you want to run only one test, you can write **f**describe instead of describe in that test

File: ddp-workspace/projects/ddp-sdk/src/public-api.ts (hunk 19 above)
Comment (MocanaAtBroad) [reply ddp-workspace/projects/ddp-sdk/src/public-api.ts]: Though as you mentioned, the formatter directive is not there and appears to work OK. So now I am confused about this public-api.ts thing and will have to do additional reading. I thought you had to import module no matter what if you want to use stuff.
I am going to adjust this anyways just to export the formatter directive as that is the only one we are using outside and will figure out what magic is going on here.

f9bd1ddddee14b34502f2cc5d546eab80e72da38

Comment (fat23cat) [reply ddp-workspace/projects/ddp-sdk/src/public-api.ts]: Ah, no problems :) 
We also use `LazyLoadResourcesDirective` outside of SDK.

**Review by fat23cat — DISMISSED** [review]

**Review by fat23cat — APPROVED** [review]
